File: .NET/Microsoft.Recognizers.Text.DateTime/German/Parsers/GermanDurationParserConfiguration.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Immutable;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Definitions.German;
using Microsoft.Recognizers.Text.Number;

namespace Microsoft.Recognizers.Text.DateTime.German
{
    public class GermanDurationParserConfiguration : BaseDateTimeOptionsConfiguration, IDurationParserConfiguration
    {
        public GermanDurationParserConfiguration(ICommonDateTimeParserConfiguration config)
            : base(config)
        {

            var numOptions = ((BaseNumberExtractor)config.CardinalExtractor).Options;
            var numConfig = new BaseNumberOptionsConfiguration(Text.Culture.German, numOptions, NumberMode.PureNumber);

            CardinalExtractor = Number.German.NumberExtractor.GetInstance(numConfig);
            NumberParser = config.NumberParser;

            DurationExtractor = new BaseDurationExtractor(new GermanDurationExtractorConfiguration(this), false);

            NumberCombinedWithUnit = GermanDurationExtractorConfiguration.NumberCombinedWithDurationUnit;

            AnUnitRegex = GermanDurationExtractorConfiguration.AnUnitRegex;
            DuringRegex = GermanDurationExtractorConfiguration.DuringRegex;
            AllDateUnitRegex = GermanDurationExtractorConfiguration.AllRegex;
            HalfDateUnitRegex = GermanDurationExtractorConfiguration.HalfRegex;
            SuffixAndRegex = GermanDurationExtractorConfiguration.SuffixAndRegex;
            FollowedUnit = GermanDurationExtractorConfiguration.DurationFollowedUnit;
            ConjunctionRegex = GermanDurationExtractorConfiguration.ConjunctionRegex;
            InexactNumberRegex = GermanDurationExtractorConfiguration.InexactNumberRegex;
            InexactNumberUnitRegex = GermanDurationExtractorConfiguration.InexactNumberUnitRegex;
            DurationUnitRegex = GermanDurationExtractorConfiguration.DurationUnitRegex;
            SpecialNumberUnitRegex = GermanDurationExtractorConfiguration.SpecialNumberUnitRegex;

            UnitMap = config.UnitMap;
            UnitValueMap = config.UnitValueMap;
            DoubleNumbers = config.DoubleNumbers;
        }

        public IExtractor CardinalExtractor { get; }

        public IDateTimeExtractor DurationExtractor { get; }

        public IParser NumberParser { get; }

        public Regex NumberCombinedWithUnit { get; }

        public Regex AnUnitRegex { get; }

        public Regex PrefixArticleRegex { get; } = null;

        public Regex DuringRegex { get; }

        public Regex AllDateUnitRegex { get; }

        public Regex HalfDateUnitRegex { get; }

        public Regex SuffixAndRegex { get; }

        public Regex FollowedUnit { get; }

        public Regex ConjunctionRegex { get; }

        public Regex InexactNumberRegex { get; }

        public Regex InexactNumberUnitRegex { get; }

        public Regex DurationUnitRegex { get; }

        public Regex SpecialNumberUnitRegex { get; }

        bool IDurationParserConfiguration.CheckBothBeforeAfter => DateTimeDefinitions.CheckBothBeforeAfter;

        public IImmutableDictionary<string, string> UnitMap { get; }

        public IImmutableDictionary<string, long> UnitValueMap { get; }

        public IImmutableDictionary<string, double> DoubleNumbers { get; }
    }
}

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/german/parsers/GermanDurationParserConfiguration.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.text.datetime.german.parsers;

import com.google.common.collect.ImmutableMap;
import com.microsoft.recognizers.text.IExtractor;
import com.microsoft.recognizers.text.IParser;
import com.microsoft.recognizers.text.datetime.config.BaseOptionsConfiguration;
import com.microsoft.recognizers.text.datetime.extractors.BaseDurationExtractor;
import com.microsoft.recognizers.text.datetime.german.extractors.GermanDurationExtractorConfiguration;
import com.microsoft.recognizers.text.datetime.parsers.config.ICommonDateTimeParserConfiguration;
import com.microsoft.recognizers.text.datetime.parsers.config.IDurationParserConfiguration;
import java.util.regex.Pattern;

public class GermanDurationParserConfiguration extends BaseOptionsConfiguration implements IDurationParserConfiguration {

    private final IExtractor cardinalExtractor;
    private final IExtractor durationExtractor;
    private final IParser numberParser;

    private final Pattern numberCombinedWithUnit;
    private final Pattern anUnitRegex;
    private final Pattern duringRegex;
    private final Pattern allDateUnitRegex;
    private final Pattern halfDateUnitRegex;
    private final Pattern suffixAndRegex;
    private final Pattern followedUnit;
    private final Pattern conjunctionRegex;
    private final Pattern inexactNumberRegex;
    private final Pattern inexactNumberUnitRegex;
    private final Pattern durationUnitRegex;
    private final Pattern specialNumberUnitRegex;

    private final ImmutableMap<String, String> unitMap;
    private final ImmutableMap<String, Long> unitValueMap;
    private final ImmutableMap<String, Double> doubleNumbers;

    public GermanDurationParserConfiguration(ICommonDateTimeParserConfiguration config) {

        super(config.getOptions());

        cardinalExtractor = config.getCardinalExtractor();
        numberParser = config.getNumberParser();
        durationExtractor = new BaseDurationExtractor(new GermanDurationExtractorConfiguration(), false);
        numberCombinedWithUnit = GermanDurationExtractorConfiguration.NumberCombinedWithDurationUnit;

        anUnitRegex = GermanDurationExtractorConfiguration.AnUnitRegex;
        duringRegex = GermanDurationExtractorConfiguration.DuringRegex;
        allDateUnitRegex = GermanDurationExtractorConfiguration.AllRegex;
        halfDateUnitRegex = GermanDurationExtractorConfiguration.HalfRegex;
        suffixAndRegex = GermanDurationExtractorConfiguration.SuffixAndRegex;
        followedUnit = GermanDurationExtractorConfiguration.DurationFollowedUnit;
        conjunctionRegex = GermanDurationExtractorConfiguration.ConjunctionRegex;
        inexactNumberRegex = GermanDurationExtractorConfiguration.InexactNumberRegex;
        inexactNumberUnitRegex = GermanDurationExtractorConfiguration.InexactNumberUnitRegex;
        durationUnitRegex = GermanDurationExtractorConfiguration.DurationUnitRegex;
        specialNumberUnitRegex = GermanDurationExtractorConfiguration.SpecialNumberUnitRegex;

        unitMap = config.getUnitMap();
        unitValueMap = config.getUnitValueMap();
        doubleNumbers = config.getDoubleNumbers();
    }

    @Override
    public IExtractor getCardinalExtractor() {
        return cardinalExtractor;
    }

    @Override
    public IExtractor getDurationExtractor() {
        return durationExtractor;
    }

    @Override
    public IParser getNumberParser() {
        return numberParser;
    }

    @Override
    public Pattern getNumberCombinedWithUnit() {
        return numberCombinedWithUnit;
    }

    @Override
    public Pattern getAnUnitRegex() {
        return anUnitRegex;
    }

    @Override
    public Pattern getDuringRegex() {
        return duringRegex;
    }

    @Override
    public Pattern getAllDateUnitRegex() {
        return allDateUnitRegex;
    }

    @Override
    public Pattern getHalfDateUnitRegex() {
        return halfDateUnitRegex;
    }

    @Override
    public Pattern getSuffixAndRegex() {
        return suffixAndRegex;
    }

    @Override
    public Pattern getFollowedUnit() {
        return followedUnit;
    }

    @Override
    public Pattern getConjunctionRegex() {
        return conjunctionRegex;
    }

    @Override
    public Pattern getInexactNumberRegex() {
        return inexactNumberRegex;
    }

    @Override
    public Pattern getInexactNumberUnitRegex() {
        return inexactNumberUnitRegex;
    }

    @Override
    public Pattern getDurationUnitRegex() {
        return durationUnitRegex;
    }

    @Override
    public Pattern getSpecialNumberUnitRegex() {
        return specialNumberUnitRegex;
    }

    @Override
    public ImmutableMap<String, String> getUnitMap() {
        return unitMap;
    }

    @Override
    public ImmutableMap<String, Long> getUnitValueMap() {
        return unitValueMap;
    }

    @Override
    public ImmutableMap<String, Double> getDoubleNumbers() {
        return doubleNumbers;
    }
}
